src/r_builtins/paste.rs

```diff
@@ -0,0 +1,360 @@
+use crate::ast::*;
+use crate::error::*;
+use crate::lang::*;
+use crate::r_vector::vectors::*;
+
+pub fn primitive_paste(args: ExprList, env: &mut Environment) -> EvalResult {
+    // Need to make sure that collapse is specified since we can not assign a
+    // default value as with sep
+    let collapse_is_specified = &args
+        .keys
+        .iter()
+        .position(|k| k == &Some("collapse".to_string()));
+
+    let R::List(vals) = env.eval_list(args)? else {
+        unreachable!()
+    };
+
+    let mut stack_vals: Vec<(_, _)> = vec![];
+    let mut sep = " ".to_string();
+    let mut collapse = String::new();
+
+    for (k, v) in &vals {
+        let k_clone = k.clone().unwrap_or("".to_string());
+        let _sep_string = String::from("sep");
+
+        match k_clone.as_str() {
+            "sep" => {
+                sep = match v {
+                    R::Vector(Vector::Character(s_v)) => s_v.get(0).unwrap().clone().to_string(),
+                    _ => {
+                        return Err(RSignal::Error(RError::Other(
+                            "sep parameter must be a character string!".to_string(),
+                        )))
+                    }
+                }
+            }
+            "collapse" => {
+                collapse = match v {
+                    R::Vector(Vector::Character(s_v)) => s_v.get(0).unwrap().clone().to_string(),
+                    _ => {
+                        return Err(RSignal::Error(RError::Other(
+                            "collapse parameter must be a character string!".to_string(),
+                        )))
+                    }
```

Comment:
Instead of operating on the raw internals, working with the higher-level R interface will help to perform coercion between data types for you so that you don't need to try to handle all sorts of edge cases. (untested code, but hopefully enough to set you off in the right direction)

```suggestion
                    R::Vector(v) => R::Vector(v.as_character()),
                    R::Null => R::Null,  // note this is a valid value for collapse
                    _ => {
                        return Err(RSignal::Error(RError::Other("invalid 'collapse' argument".to_string())))
                    }
```

Reply:
We need to check whether the supplied sep value is a character. R does not accept non-character values. If we use
`R::Vector(v) => R::Vector(v.as_character())` then this will coerce non-valid sep value to a character.

Reply:
In the future I think it would be nice to allow anything that can be coerced to character, but let's start with matching R. This looks good.